Calculating RDMs

HeikoSchuett opened the pull request:
This will be the code to calculate RDMs from datasets

caiw commented:
This appears to be a duplicate of #28, so I'm closing it. Please let me know if I've done this in error!

yunshiuan commented:
Sorry I was just uncertain if the pull request should be aimed at the develop branch or the master one or both.

caiw commented:
No problem. I think we put non-emergency fixes and improvements onto `develop`, and will merge into `master` for a larger release.

HeikoSchuett committed:
copied RDM calculation from local code to start editing

HeikoSchuett committed:
implemented contrast matrix

HeikoSchuett committed:
averaging dataset

HeikoSchuett committed:
added averaging of datasets

HeikoSchuett committed:
dataset computation improved

HeikoSchuett committed:
addded documentation to rdm utils

HeikoSchuett committed:
added pattern_descriptors to RDM class

HeikoSchuett committed:
first version of calc rdm euclid

HeikoSchuett committed:
exposed calc_rdm and first test of euclidean case

HeikoSchuett committed:
Mahalanobis distance implemented

HeikoSchuett committed:
allowed subset with multiple values in list

HeikoSchuett committed:
first crossnobis implementation

HeikoSchuett committed:
style updates

HeikoSchuett committed:
style updates

HeikoSchuett committed:
style updates

HeikoSchuett commented:
I'll have a look at the refactoring suggestions of codeclimate and implement the missing cosine distance.
Then the only remaining issue is the stability of the crossnobis ram to imbalanced designs

jdiedrichsen reviewed (commented):
@ Heiko: The functionality of providing a contrast matrix is already done in util.indicator.allpairs 
(all pairwise contrasts). 
It's slightly more general, but util.indicator.allpairs(range(0,n_patterns)) does the same job, I think
Is there a reason why we need this extra version?

HeikoSchuett committed:
removed superflous contrast_matrix

HeikoSchuett commented:
No, there is no reason to double it. I just removed it. I was simply not aware of all pairs existence when I copied the code from my code here

HeikoSchuett committed:
refactoring to include bool_index for descriptors

HeikoSchuett committed:
refactoring sugested by code climate

HeikoSchuett committed:
more refactoring for code climate

HeikoSchuett committed:
style fixes

HeikoSchuett committed:
removed checks for RDM from square matrix to avoid rounding problems

HeikoSchuett committed:
added correlation distance

HeikoSchuett committed:
style fix

HeikoSchuett commented:
I would now suggest to merge this.
This does not do all variants of RDMs we might want, but we can add more later and I have issues on them open already. 

So: Please review and raise points where I need to change things still

HeikoSchuett commented:
closes #25

JasperVanDenBosch commented on pyrsa/rdm/calc_rdm.py:
```diff
@@ -0,0 +1,200 @@
+                obs_descriptor used to define the rows/columns of the RDM
+                defaults to one row/column per row in the dataset
+            noise (numpy.ndarray):
+                precision matrix used to calculate the RDM
```
Could you add something about the expected dimensions of the noise array?

HeikoSchuett replied:
I added a line to the help that this is supposed to have dimension n_channel x n_channel and introduced corresponding checks in a new commit

HeikoSchuett committed:
added noise dimension checks

HeikoSchuett committed:
merge master

JasperVanDenBosch commented:
unit test in progress.. will commit tomorrow

JasperVanDenBosch committed:
unit test proposal for calc_euclid

JasperVanDenBosch commented:
@HeikoSchuett the import `from pyrsa.rdm.calc_rdm import calc_rdm` in `pyrsa.rdm.__init__.py` currently 'hides' the `calc_rdm` module (as they have the same name). This means it's not possible to import he module itself anymore I think. Should we rename the module? Or the import shortcut?

HeikoSchuett committed:
rename calc_rdm module into calc and adjusted test of euclidean distance rdm

HeikoSchuett commented:
> @HeikoSchuett the import `from pyrsa.rdm.calc_rdm import calc_rdm` in `pyrsa.rdm.__init__.py` currently 'hides' the `calc_rdm` module (as they have the same name). This means it's not possible to import he module itself anymore I think. Should we rename the module? Or the import shortcut?

To fix this issue I now renamed the submodule into `calc`. Thank you for finding this!

I also changed the test you wrote to compare to the squared Euclidean distance divided by the number of channels as this is what the calc_rdm_euclid computes. It now passes.

JasperVanDenBosch commented:
Great. I think this could serve as a good example of how to make a unit test that actually tests the main objective of the unit. Also it only tests the specific unit by mocking out the other class involved. (If dataset has a bug, it should not be the rdm tests that fail). The main benefit of this is that it shows the dependencies between the two classes clearly.

I'm ok with merging this now and then adding more tests latsr

HeikoSchuett committed:
allowing np.arrays as subset values

HeikoSchuett committed:
style fix

HeikoSchuett committed:
style fix

HeikoSchuett committed:
merge master
